File: core/src/main/scala/tajo/benchmark/BenchmarkSet.scala
package tajo.benchmark

import tajo.client.TajoClient
import org.apache.hadoop.net.NetUtils
import collection.immutable.HashMap
import java.io.File
import tajo.catalog.{TConstants, Schema}
import tajo.util.FileUtil
import tajo.catalog.store.MemStore
import tajo.conf.TajoConf
import tajo.conf.TajoConf.ConfVars

/**
 * @author Hyunsik Choi
 */

abstract class BenchmarkSet() {
  protected var tajo : TajoClient = _
  protected var schemas = HashMap[String, Schema]()
  protected var outSchemas = HashMap[String, Schema]()
  protected var datadir : String = _

  def init(conf : TajoConf, _datadir : String) {
    this.datadir = _datadir
    if (System.getProperty(ConfVars.MASTER_ADDRESS.varname) != null) {
      tajo = new TajoClient(NetUtils.createSocketAddr(
        System.getProperty(ConfVars.MASTER_ADDRESS.varname)))
    } else {
      conf.set(TConstants.STORE_CLASS, classOf[MemStore].getCanonicalName)
      tajo = new TajoClient(conf)
    }
  }

  protected var queries = HashMap[String, String]()

  protected def loadQueries(dir : String) {
    val queryDir = new File(dir)
    val files = queryDir.list()
    for (file <- files if file.endsWith(".tql")) {
      val last = file.indexOf(".tql")
      val name = file.substring(0, last)
      val query = FileUtil.readTextFile(new File(queryDir + "/" + file))
      queries += (name -> query)
    }
  }

  def loadSchemas()

  def loadOutSchema()

  def loadQueries()

  def loadTables()

  def getTableNames : Array[String] = {
    schemas.keySet.toArray
  }

  def getQuery(queryName : String) : String = {
    queries(queryName)
  }

  def getSchema(name : String) : Schema = {
    schemas(name)
  }

  def getSchemas() : Iterator[Schema] = {
    schemas.valuesIterator
  }

  def getOutSchema(name : String) : Schema = {
    outSchemas(name)
  }

  def perform(queryName : String) {
    val query = queries(queryName)
    if (query == null) {
      throw new IllegalArgumentException("#{queryName} does not exists")
    }
    val start = System.currentTimeMillis()
    tajo.executeQuery(query)
    val end = System.currentTimeMillis()

    println("====================================")
    println("QueryName: " + queryName)
    println("Query: " + query)
    println("Processing Time: " + (end - start))
    println("====================================")
  }
}

File: tajo-core/tajo-core-backend/src/main/java/tajo/benchmark/BenchmarkSet.java
package tajo.benchmark;

import com.google.protobuf.ServiceException;
import org.apache.hadoop.net.NetUtils;
import tajo.catalog.CatalogConstants;
import tajo.catalog.Schema;
import tajo.catalog.store.MemStore;
import tajo.client.TajoClient;
import tajo.conf.TajoConf;
import tajo.conf.TajoConf.ConfVars;
import tajo.util.FileUtil;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public abstract class BenchmarkSet {
  protected TajoClient tajo;
  protected Map<String, Schema> schemas = new HashMap<String, Schema>();
  protected Map<String, Schema> outSchemas = new HashMap<String, Schema>();
  protected Map<String, String> queries = new HashMap<String, String>();
  protected String dataDir;

  public void init(TajoConf conf, String dataDir) throws IOException {
    this.dataDir = dataDir;
    if (System.getProperty(ConfVars.TASKRUNNER_LISTENER_ADDRESS.varname) != null) {
      tajo = new TajoClient(NetUtils.createSocketAddr(
          System.getProperty(ConfVars.TASKRUNNER_LISTENER_ADDRESS.varname)));
    } else {
      conf.set(CatalogConstants.STORE_CLASS, MemStore.class.getCanonicalName());
      tajo = new TajoClient(conf);
    }
  }

  protected void loadQueries(String dir) throws IOException {
    File queryDir = new File(dir);

    int last;
    String name, query;
    for (String file : queryDir.list()) {
      if (file.endsWith(".tql")) {
        last = file.indexOf(".tql");
        name = file.substring(0, last);
        query = FileUtil.readTextFile(new File(queryDir + "/" + file));
      }
    }
  }

  public abstract void loadSchemas();

  public abstract void loadOutSchema();

  public abstract void loadQueries() throws IOException;

  public abstract void loadTables() throws ServiceException;

  public String [] getTableNames() {
    return schemas.keySet().toArray(new String[schemas.size()]);
  }

  public String getQuery(String queryName) {
    return queries.get(queryName);
  }

  public Schema getSchema(String name) {
    return schemas.get(name);
  }

  public Iterable<Schema> getSchemas() {
    return schemas.values();
  }

  public Schema getOutSchema(String name) {
    return outSchemas.get(name);
  }

  public void perform(String queryName) throws IOException, ServiceException {
    String query = getQuery(queryName);
    if (query == null) {
      throw new IllegalArgumentException("#{queryName} does not exists");
    }
    long start = System.currentTimeMillis();
    tajo.executeQuery(query);
    long end = System.currentTimeMillis();

    System.out.println("====================================");
    System.out.println("QueryName: " + queryName);
    System.out.println("Query: " + query);
    System.out.println("Processing Time: " + (end - start));
    System.out.println("====================================");
  }
}
